### PR TITLE
An alternative implementation of general finite mixture models

### PR DESCRIPTION
**Pull request by dfm**
For a project I needed a more general Mixture* distribution allowing for component distributions from different families. Since #1413 wasn't ready for use, I implemented my own version that is somewhat more complete and consistent with the existing `MixtureSameFamily` distribution. In case it would be of interest to this project, I figured I'd open this here. If the maintainers would like such a distribution be included, I'm happy to add some tests and respond to any feedback. If it's not of interest (with a preference for #1413 or otherwise), absolutely no worries and I'm happy to close this. Thanks!!

### CODE REVIEW

**Comment by hessammehr** [comment]
Hi @dfm, author of #1413 here. Not a maintainer but comparing the two implementations, I agree that yours does seems more complete and compatible with the existing same family mixture implementation. Personally, I would be in favour of proceeding with this PR instead of #1403.

File: numpyro/distributions/mixtures.py [hunk 1]
```diff
@@ -211,3 +223,208 @@ def log_prob(self, value, intermediates=None):
         component_log_probs = self.component_distribution.log_prob(value)
         sum_log_probs = self.mixing_distribution.logits + component_log_probs
         return jax.nn.logsumexp(sum_log_probs, axis=-1)
+
+
+class MixtureGeneral(Distribution):
+    def __init__(
+        self, mixing_distribution, distributions, *, validate_args=None
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:230]: nit: maybe `distributions` -> `component_distributions` to match MixtureSameFamily api

File: numpyro/distributions/mixtures.py [hunk 2]
```diff
@@ -211,3 +223,208 @@ def log_prob(self, value, intermediates=None):
...
+    ):
+        _check_mixing_distribution(mixing_distribution)
+
+        self._mixture_size = jnp.shape(mixing_distribution.probs)[-1]
+        if isinstance(distributions, Distribution):
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:235]: Hmm, do we need to support mixtures of the same distributions?

File: docs/source/distributions.rst [hunk 3]
```diff
@@ -540,6 +540,13 @@ MixtureSameFamily
     :show-inheritance:
     :member-order: bysource
 
+MixtureGeneral
```
Comment (fehiepsi) [inline docs/source/distributions.rst:543]: Maybe also `Mixture` (not autoclass but `.. autofunction::`)

**Review by fehiepsi — COMMENTED** [review]
Thanks @dfm! This is a nice distribution to have. Given that @hessammehr would like to close that PR (thanks @hessammehr for the initial take), I think we can move ahead with this one. :)

File: numpyro/distributions/mixtures.py [hunk 4]
```diff
@@ -10,6 +10,21 @@
 from numpyro.distributions.util import is_prng_key, validate_sample
 
 
+def Mixture(mixing_distribution, distributions, *, validate_args=None):
+    """
+    Mixture distribution.
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:15]: Could you add an example section for this as in `MixtureSameFamily`?

**Comment by dfm** [comment]
Thanks for the comments @fehiepsi! I'll incorporate the comments and add some docs and tests ASAP.

File: numpyro/distributions/mixtures.py [hunk 5]
```diff
@@ -211,3 +223,208 @@ def log_prob(self, value, intermediates=None):
...
+                    "numpyro.distributions.Distribution subclasses"
+                )
+        if len(distributions) != self.mixture_size:
+            raise ValueError(
+                "The number of elements in 'distributions' must match the mixture size; "
+                f"expected {self._mixture_size}, got {len(distributions)}"
+            )
+
+        # FIXME: It would be good to check the support, but __eq__ isn't implemented
+        #        properly for most constraints...
```
[the remark this answers is not in the document]
Comment (dfm) [reply numpyro/distributions/mixtures.py]: Done!

File: numpyro/distributions/mixtures.py [hunk 6]
```diff
@@ -211,3 +223,208 @@ def log_prob(self, value, intermediates=None):
...
+        #     raise ValueError("All distributions must have the same support.")
+
+        self._mixing_distribution = mixing_distribution
+        self._distributions = distributions
+
+        batch_shape = lax.broadcast_shapes(
+            mixing_distribution.batch_shape,
+            *(d.batch_shape for d in distributions),
+        )
+        event_shape = lax.broadcast_shapes(
```
[the remark this answers is not in the document]
Comment (dfm) [reply numpyro/distributions/mixtures.py]: Good point - done.

File: numpyro/distributions/mixtures.py (hunk 1 above)
Comment (dfm) [reply numpyro/distributions/mixtures.py]: 👍

File: numpyro/distributions/mixtures.py (hunk 2 above)
Comment (dfm) [reply numpyro/distributions/mixtures.py]: Fair - we handle this use case in `Mixture`.

**Comment by dfm** [comment]
@fehiepsi — I'd addressed your comments, added some docstrings, and tests for this new distribution.

Also since you last looked I refactored much of the common implementation details for `MixtureSameFamily` and `MixtureGeneral` into a common abstract base class. This definitely simplifies the implementation without changing behavior, so I think it's probably a good choice, but always happy for more feedback!

**Review by fehiepsi — COMMENTED** [review]
Hi @dfm, could you merge master branch to make CI pass?

File: numpyro/distributions/mixtures.py [hunk 7]
```diff
@@ -10,19 +10,166 @@
...
 
-    As an example:
+    and more details can be found in the documentation for each of these
+    classes.
+
+    :param mixing_distribution: A :class:`~numpyro.distributions.Categorical`
+        specifying the weights for each mixture components. The size of this
+        distribution specifies the number of components in the mixture,
+        ``mixture_size``.
+    :param distributions: Either a list of component distributions or a single
```
[the remark this answers is not in the document]
Comment (dfm) [reply numpyro/distributions/mixtures.py]: Good catch - thanks!!

**Comment by dfm** [comment]
Pulled from `master` and tests passing locally.

**Review by fehiepsi — APPROVED** [review]
Wow a lot of tricky design! Thanks a lot, @dfm!

**Comment by dfm** [comment]
@fehiepsi — Thanks for your review and all your work on this project!!

**Pull request merged**